src/Microsoft.AspNet.Mvc.Core/SessionStateTempDataProvider.cs

```diff
@@ -37,7 +65,27 @@ public class SessionStateTempDataProvider : ITempDataProvider
                 using (var memoryStream = new MemoryStream(value))
                 using (var writer = new BsonReader(memoryStream))
                 {
-                    tempDataDictionary = jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                    tempDataDictionary = _jsonSerializer.Deserialize<Dictionary<string, object>>(writer);
+                }
+                foreach (var item in tempDataDictionary.ToList())
```

Comment:
Could you create a new dictionary rather than writing to this one (preferably one with `StringComparer.OrdinalIgnoreCase` semantics? Seems cleaner than doing a `ToList` here.